### PR TITLE
Improve predict / evaluate

### LINKED ISSUE #82: Train should only train not evaluate
At the moment train also evaluates but since we have evaluate, it would be better to get the workflow to run train and then evaluate. This also ensure that evaluates runs correctly.

### LINKED ISSUE #104: Improve predict / evaluate
Evaluate at this point happens in two places `train_and_evaluate` in `train.py` and `evaluate_model` in `evaluate_model.py`. We should remove the evaluate part from train and starting using the `evaluate_model`. We should also update configs where necessary to make use of evaluate along with params needed.

Both evaluate and predict need access to the model and currently this is provided by a custom load model functionality inside `predict` in `predict.py`. We should instead leverage `create_model` and try to use `.load(model_path)` and if it does not exist to fallback to using pickling or test if the path is a pickle and otherwise use `.load()`.

To be fair we were using create_model before but since all models were defined where `create_model` lived every time we introduced a new model, everything needed to be updated. Since we decoupled the models and defined custom models that implement a `load` function it is now easier to revert to using create_model as the means of instating the model before loading.

In the future we might make better use of `meta.json` that some models have to detect and load the right model.

This is also related to #82

### LINKED ISSUE #115: Split train / test in separate script
Currently data are being split for `train/test` in `grants_tagger/utils.py` in `load_train_test_data`. This means that when we evaluate or want to retrieve the test set we need to use the same function and ensure we run it the same way which is not ideal and it lead to errors.

Instead we should split the data into a train / test set after preprocess and before train and use this split throughout.

Some considerations
* preprocess_mesh currently splits train and test in the absence of a separate script. the functionality needs to be removed from there
* evaluate_model and potentially other places use `load_train_test_data` to get a test set with the same split as in train. this should be removed together with the flag split data or the flag should be turned to false.
* `grants_tagger/utils.py` contains logic on how to split, this should be removed and moved to a separate file, in theory it should only contain logic how to load data in memory or via a generator

### PR DESCRIPTION
**Pull request by nsorros**
## Description

This PR resolves a small technical debt that piled up in `predict` and `evaluate` which was that models were being initialised in a custom way inside `predict` instead of using `create_model` which is the "de facto" way of initialising models in grants_tagger.

Additionally `evaluate_model` was not being used during training as `train` had its own evaluation. We are resolving this here and making `evaluate` the only way to evaluate models. This is already the case for production models.

Fixes #104
Fixes #115 
Fixes #82 

## Checklist

- [x] Linked to Notion or GitHub issue
- [x] Added tests
- [x] Updated README
- [ ] DVC up to date

### CODE REVIEW

**Review by aCampello — APPROVED** [review]
Very nice PR, always good to clear out technical debt when there is time to do so.

I put some comments about comments, mainly for my understanding.

File: grants_tagger/evaluate_model.py [hunk 1]
```diff
@@ -1,47 +1,71 @@
...
+
+    if type(threshold) != list:
+        threshold = [threshold]
 
+    widths = (12, 5, 5, 5)
+    header = ["Threshold", "P", "R", "F1"]
+    print(table([], header, divider=True, widths=widths))
+
+    results = []
+    for th in threshold:
```
[the remark this answers is not in the document]
Comment (nsorros) [reply grants_tagger/evaluate_model.py:55]: Haha do we want that? I removed the classification report because it produces one line per label so 30K lines for mesh

File: tests/test_predict.py [hunk 2]
```diff
@@ -8,8 +8,13 @@
...
 
-from grants_tagger.train import create_label_binarizer, train_and_evaluate
-from grants_tagger.predict import predict_tags, predict
+from grants_tagger.train import create_label_binarizer, train
+from grants_tagger.predict import predict_tags
+try:
+    from grants_tagger.models.mesh_xlinear import MeshXLinear
+    MESH_XLINEAR_IMPORTED = True
+except ImportError:
+    MESH_XLINEAR_IMPORTED = False
```
[the remark this answers is not in the document]
Comment (nsorros) [reply tests/test_predict.py:17]: This is a test

File: grants_tagger/evaluate_model.py [hunk 3]
```diff
@@ -1,47 +1,71 @@
...
+Evaluate model performance on test set
 """
 import pickle
 import json
 
-from sklearn.model_selection import train_test_split
-from sklearn.metrics import classification_report, precision_recall_fscore_support
-from wasabi import table
+from sklearn.metrics import precision_recall_fscore_support
+from wasabi import table, row
```
[the remark this answers is not in the document]
Comment (nsorros) [reply grants_tagger/evaluate_model.py:8]: 😬

File: grants_tagger/evaluate_model.py [hunk 4]
```diff
@@ -1,47 +1,71 @@
...
+import scipy.sparse as sp
 
 from grants_tagger.utils import load_train_test_data, load_data
-from grants_tagger.predict import predict
+from grants_tagger.models.create_model import load_model
+
+
+# TODO: Move inside model that need to produce sparse probs
```
[the remark this answers is not in the document]
Comment (nsorros) [reply grants_tagger/evaluate_model.py:15]: I added an issue https://github.com/wellcometrust/grants_tagger/issues/113 so removing the comment

File: grants_tagger/models/create_model.py [hunk 5]
```diff
@@ -231,3 +216,14 @@ def create_model(approach, parameters=None):
...
+
+
+def load_model(approach, model_path):
+    if str(model_path).endswith(".pkl"):
+        model = load_pickle(model_path)
+    else:
+        model = create_model(approach)
+        model.load(model_path)
+
+    # The problem becomes the ensemble
```
[the remark this answers is not in the document]
Comment (nsorros) [reply grants_tagger/models/create_model.py:228]: it was note for me, removing

File: grants_tagger/models/mesh_cnn.py [hunk 6]
```diff
@@ -165,14 +165,14 @@ def predict(self, X):
         return Y_pred
 
     def predict_proba(self, X):
-        # TODO: Maybe not needed beacause CNN batches in predict. Test in disease MeSH.
         if type(X) in [list, np.ndarray]:
             X_vec = self.vectorizer.transform(X)
+            # TODO: Move to CNNClassifier with sparse_y flag or remove sparse_y argument
```
[the remark this answers is not in the document]
Comment (nsorros) [reply grants_tagger/models/mesh_cnn.py:170]: https://github.com/wellcometrust/grants_tagger/issues/134

File: grants_tagger/models/create_model.py [hunk 7]
```diff
@@ -205,22 +205,7 @@ def create_model(approach, parameters=None):
...
     elif approach == 'science-ensemble':
-        tfidf_svm = Pipeline([
-            ('tfidf', TfidfVectorizer(
-                stop_words='english', max_df=0.95,
-                min_df=0.0, ngram_range=(1,1)
-            )),
-            ('svm', OneVsRestClassifier(SVC(kernel='linear', probability=True)))
-        ])
-        scibert = BertClassifier(pretrained="scibert") 
-        model = WellcomeVotingClassifier(
```
[the remark this answers is not in the document]
Comment (nsorros) [reply grants_tagger/models/create_model.py:216]: Note that this is replaced and part of `ScienceEnsemble`

**Comment by nsorros** [comment]
This is ready to merge, waiting to get an instance to run `dvc repro`
